### DISCUSSION TITLE
*##[src^="data:"] in My filters has no effect with easylist and/or fanboy-social

### DISCUSSION THREAD

**Discussion by emanruse** [uBlockOrigin/uAssets#22689] [General]
### Prerequisites

- [X] This is **NOT** a [_YouTube_](https://www.github.com/uBlockOrigin/uAssets/issues/20586), [_Facebook_](https://www.github.com/uBlockOrigin/uAssets/issues/3367) or [_Twitch_](https://www.github.com/uBlockOrigin/uAssets/issues/5184) report. _These sites MUST be reported by clicking their **respective links**_.
- [X] I read and understand the [policy about what is a valid filter issue](https://github.com/uBlockOrigin/uAssets/blob/master/README.md#uassets).
- [X] I verified that this issue is not a duplicate. (Search [here](https://github.com/uBlockOrigin/uAssets/issues) to find out.)
- [X] I did not remove any of the [default filter lists](https://github.com/uBlockOrigin/uAssets/assets/66517106/6e8d024a-5869-47cf-b142-b2357eb6c7d0), or I have verified that the issue was not caused by removing any of the default lists.
- [X] I did not enable additional filter lists, or I have verified that the issue still occurs without enabling additional filter lists.
- [X] I do not have custom filters/rules, or I have verified that the issue still occurs without custom filters/rules.
- [X] I am not using uBlock Origin along with other content blockers.
- [X] I have verified that the web browser's built-in blocker or [DNS blocking](https://en.wikipedia.org/wiki/DNS_blocking) (standalone or through a VPN) is not causing the issue.
- [X] I have verified that other extensions are not causing the issue.
- [X] If this is about a breakage or detection, I have verified that it is caused by uBlock Origin and isn't a site issue.
- [ ] I did not answer truthfully to **ALL** the above checkboxes.

### URL(s) where the issue occurs.

```yaml
https://github.com/404
```

##### Disclaimer:

I am not 100% sure if this is correct to be reported as filter list issue or as an issue with uBO itself, in the sense that My Filters obviously has lower priority than external filter lists. So, for the sake of completeness, I will report it as both, because the documentation does not clarify which static filter has priority over another.

1. Use the following settings:

```
{
  "version": "1.56.0",
  "userSettings": {
    "firewallPaneMinimized": false,
    "importedLists": []
  },
  "selectedFilterLists": [
    "user-filters"
  ],
  "hiddenSettings": {},
  "whitelist": [
    "about-scheme",
    "moz-extension-scheme"
  ],
  "dynamicFilteringString": "* * * block\n* * inline-script block\n* * 1p-script block\n* * 3p-script block\n* * 3p-frame block\n* * image block\n* * 3p block\nbehind-the-scene * * block",
  "urlFilteringString": "",
  "hostnameSwitchesString": "no-popups: * true\nno-remote-fonts: * true\nno-large-media: * true\nno-csp-reports: * true\nno-scripting: * true",
  "userFilters": "*$stylesheet\n*$document,csp=style-src\n*##[src^=\"data:\"]\n"
}
```

2. Navigate to https://github.com/404

**Expected:** The `*##[src^="data:"]` in My Fiilters should hide the base64 images.

**Actual:** As expected.

3. Enable EasyList (easylist) or EasyList – Social Widgets (fanboy-social) filter lists. (or both)

4. Refresh the page (with disabled cache in web console).

**Result:** `*##[src^="data:"]` has no effect. base64 images are displayed.

Tested each and every other list (with all other lists disabled). Only the 2 lists mentioned above cause this effect.

### Other extensions used

none

### Screenshot(s)

I cannot enable JavaScript for privacy reasons. The STR is detailed and simple enough to reproduce and the result is consistent.

### Configuration

<details>
<summary>Details</summary>

```yaml
uBlock Origin: 1.56.0
Firefox: 115
filterset (summary):
 network: 47581
 cosmetic: 34229
 scriptlet: 0
 html: 0
listset (total-discarded, last-updated):
 removed:
  ublock-filters: null
  ublock-badware: null
  ublock-privacy: null
  ublock-unbreak: null
  ublock-quick-fixes: null
  easyprivacy: null
  urlhaus-1: null
  plowe-0: null
 default:
  user-filters: 3-0, never
  easylist: 81820-13, never
filterset (user): [array of 3 redacted]
trustedset:
 removed:
  chrome-extension-scheme
  chrome-scheme
  edge-scheme
  opera-scheme
  vivaldi-scheme
  wyciwyg-scheme
switchRuleset:
 added: [array of 4 redacted]
 removed:
  no-large-media: behind-the-scene false
hostRuleset:
 added: [array of 8 redacted]
 removed:
  behind-the-scene * * noop
  behind-the-scene * image noop
  behind-the-scene * 3p noop
  behind-the-scene * inline-script noop
  behind-the-scene * 1p-script noop
  behind-the-scene * 3p-script noop
  behind-the-scene * 3p-frame noop
userSettings: [none]
hiddenSettings: [none]
supportStats:
 allReadyAfter: 276 ms
 maxAssetCacheWait: 22 ms
 cacheBackend: browser.storage.local
popupPanel:
 blocked: 0
 no-large-media: true
 no-popups: true
 no-remote-fonts: true
 no-scripting: true

```
</details>

**Comment by ghost** [comment]
Use specific filter  (`github.com##[src^="data:"]`) or override the `$ghide` (`*$ghide,important`).

<sub>https://github.com/uBlockOrigin/uBlock-issues/issues/3141</sub>

**Comment by garry-ut99** [comment]
> emanruse : the documentation does not clarify which static filter has priority over another.

It does clarify, exception filters override normal filters : 
- https://help.adblockplus.org/hc/en-us/articles/360062733293-How-to-write-filters#allowlist
- https://help.adblockplus.org/hc/en-us/articles/360062733293-How-to-write-filters#elemhide_exceptions
- https://help.adblockplus.org/hc/en-us/articles/360062733293-How-to-write-filters#generic-specific
- https://github.com/uBlockOrigin/uBlock-issues/wiki/Static-filter-syntax#specific-generic

Also some additional filter options further override exceptions :
- `badfilter` https://github.com/uBlockOrigin/uBlock-issues/wiki/Static-filter-syntax#badfilter :  
(for example: `@@||github.com^$generichide,badfilter` to disable the whitelist filter)
-  `important` https://github.com/gorhill/uBlock/issues/2904#issuecomment-325198907
-  dynamic filtering https://github.com/gorhill/uBlock/issues/2904#issuecomment-325201705

<hr>

> gwarser : Use specific filter `github.com##[src^="data:"]`

Or `com##[src^="data:"]`

**Comment by emanruse** [comment]
> It does clarify, exception filters override normal filters : 
> - https://help.adblockplus.org/hc/en-us/articles/360062733293-How-to-write-filters#allowlist
> - https://help.adblockplus.org/hc/en-us/articles/360062733293-How-to-write-filters#elemhide_exceptions

This only confirms that external filters can override user-set local ones, which makes the prioritization of "My filters" over external ones (the other issue) a valid one.

The goal of *##[src^="data:"] is to block all base64 (on all sites). The github link was just an example to illustrate the issue.

The suggested:

> `github.com##[src^="data:"]`

> Or `com##[src^="data:"]`

> `@@||github.com^$generichide,badfilter`

are limited to a domain or TLD, i.e. the goal is still impossible.


> or override the `$ghide` (`*$ghide,important`).

That works but wouldn't it have other (undesired) side effects?

The point is: How does that resolve the actual issue of external filters overriding local ones?

> -  dynamic filtering https://github.com/gorhill/uBlock/issues/2904#issuecomment-325201705

Seems unrelated to in-document base64 data.

**Comment by garry-ut99** [comment]
If the proposed solutions still don't satisfy you, try to continue here : https://github.com/uBlockOrigin/uBlock-issues/issues/3141 and argue with the owner.

**Comment by emanruse** [comment]
What do you mean "to continue"? I am the author of both issues. Both were closed as "completed" without action whatsoever while the actual issue still remains.

Quite confusing.

**Comment by garry-ut99** [comment]
This is a filter issue tracker, and we provided you already all possible filter solutions, we don't have any more, if your issue can't be solved by filters, then the only way is to your feature request be implemented in uBO, and the proper place for feature requests is uBO issue tracker, not filter issue tracker, the issue in uBO tracker is closed but you can still try to ask / convince the uBO creator.

**Comment by garry-ut99** [comment]
> `*##[src^="data:"][src^="data:"]`

It works but only for normal exceptions (`#@#[src^="data:"]`), it doesn't work for generichide.

**Comment by emanruse** [comment]
>     *##[src^="data:"][src^="data:"]

Same result - the base64 images show up.

**Comment by garry-ut99** [comment]
**1**
> emanruse :  or override the `$ghide` (`*$ghide,important`).
> emanruse : That works but wouldn't it have other (undesired) side effects?

Yes, it'll have bad side-effects unfortunately, because `generichide` exceptions are used to fool anti-adblock scripts on many websites, disabling `generichide` globally will cause anti-adblock appear on many sites.

**2**
> emanruse : The suggested: `github.com##[src^="data:"]` Or `com##[src^="data:"]`
> emanruse : are limited to a domain or TLD, i.e. the goal is still impossible.

Yes, the best what can be done at the moment is to combine several dozen of most common TLDs + `*`, 
for example according to `https://www.icdsoft.com/blog/what-are-the-most-popular-tlds-domain-extensions/` :
`com,net,org,de,icu,uk,ru,info,top,xyz,tk,cn,ga,cf,nl,*##[src^="data:"]`
or a bigger one crafted out of 100 top TLDs according to 
`https://www.scoutdns.com/100-most-popular-tlds-by-google-index/` :  
`com,net,org,jp,de,uk,fr,br,it,ru,es,me,gov,pl,ca,au,cn,co,in,nl,edu,info,eu,ch,id,at,kr,cz,mx,be,tv,se,tr,tw,al,ua,ir,vn,cl,sk,ly,cc,to,no,fi,us,pt,dk,ar,hu,tk,gr,il,news,ro,my,biz,ie,za,nz,sg,ee,th,io,xyz,pe,bg,hk,rs,lt,link,ph,club,si,site,mobi,by,cat,wiki,la,ga,xxx,cf,hr,ng,jobs,online,kz,ug,gq,ae,is,lv,pro,fm,tips,ms,sa,app,*##[src^="data:"]`

**3**
Or another alternative solution is to use HTML filter https://github.com/gorhill/uBlock/wiki/Static-filter-syntax#html-filters : `*##^[src^="data:"]`, works only in Firefox, will work on most websites, but might not work on some sites due to not supporting some rare page encodings, also will not remove DOMs created by JS during page loading or after page loaded, because it works only at source code level.

**Comment by krystian3w** [comment]
Instead wrote top 100 public suffixes or HTML filtering it is possible use untranslatable to a simple filter the injection of some CSS (yet/for now):

```adb
*##[src^="data:"]:style(display: none !important; width: unset)
```

(`width: unset` can by changed to anything like `dummy: 1` to avoid simplification filter)

[Injection globally CSS is in use to disable advertising spectrum blocking detection, so blocking the use of this may unblock some spectrum blocking modals/alerts, if we consider it a bug that needs to be corrected in the next 10 years]

Or little slower `:remove()` (blinks at smartphone).

---
---

If placeholders with borders are not bad then also works:


```adb
*##[src^="data:"]:remove-attr(src)
``` 

(Also in legacy `ra.js`)


(https://github.com/uBlockOrigin/uAssets/assets/35370833/98c80a3e-a01e-4dda-afa3-91626f22594f)

**Comment by emanruse** [comment]
> *##[src^="data:"]:style(display: none !important; width: unset)

Thanks.
This works, even without "width: unset" (what's its purpose?)

**Reply by ghost** [reply]
If I recall correctly, there was a code to treat `:style(display: none !important;)` as plain hiding filters (dropping the `:style()` altogether). Adding any other style will keep it as a styling filter. If it works without it, then maybe the purging code is in different code branch or it was changed.

**Reply by krystian3w** [reply]
Maybe this was limited to AG syntax only (used to disable see AG rules in CB for Android from Play Store, iPhones/iPads):

> `*#$#foo { display: none ! Important; }`

But AdGuard no needs fallback with "aka All URL-s" e.g. legacy concept of his JS scripts rules (before accept new scriptlets system) still is injected on whole Internet.

**Comment by emanruse** [comment]
So is 'width: unset' necessary?

**Reply by krystian3w** [reply]
For now yes, check this on pressed <kbd>Ctrl</kbd> (or Command <lbs>⌘</kbd>) + <kbd>F5</lbs> to exclude load bugged cache.

**Comment by emanruse** [comment]
I have checked reloading with cache disabled. With or without 'width: unset' the result is the same but I accept your word that it is required (though I still don't understand why).

**Reply by krystian3w** [reply]
Mainly so that the filter compiler does not simplify this into a non-functioning title filter.

**Comment by emanruse** [comment]
OK. Thanks.

**Comment by gorhill** [comment]
> there was a code to treat `:style(display: none !important;)` as plain hiding filters

This was probably before `:style()` filters could be injected declaratively.
